stan/math/prim/mat/fun/gp_matern32_cov.hpp

```diff
@@ -0,0 +1,278 @@
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ *
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern32_cov(const std::vector<T_x> &x, const T_s &sigma,
+                const T_l &length_scale) {
+  using std::exp;
+  using std::pow;
+
+  size_t x_size = x.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern32_cov", "x", x[n]);
+
+  check_positive_finite("gp_matern32_cov", "magnitude", sigma);
+  check_positive_finite("gp_matern32_cov", "length scale", length_scale);
+
+  Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_3_inv_l = sqrt(3.0) / length_scale;
+  T_l neg_root_3_inv_l = -1.0 * sqrt(3.0) / length_scale;
+
+  for (size_t i = 0; i < x_size; ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      typename return_type<T_x, T_s, T_l>::type distance
+          = sqrt(squared_distance(x[i], x[j]));
+      cov(i, j) = sigma_sq * (1.0 + root_3_inv_l * distance)
+                  * exp(neg_root_3_inv_l * distance);
+      cov(j, i) = cov(i, j);
+    }
+  }
+  return cov;
+}
+
+/**
+ * Returns a Matern 3/2 covariance matrix
+ *
+ * \f[ k(x, x') = \sigma^2(1 + \sqrt{3}
+ *   \sqrt{\sum_{k=1}^{K}\frac{d{(x, x')^2}}{l_k^2}})
+ *   exp(-\sqrt{3}\sqrt{\sum_{k=1}^{K}\frac{d(x, x')^2}{l_k^2}}) \f]
+ *
+ * where \f$d(x, x')\f$ is the Euclidean distance.
+ *
+ * @tparam T_x type of elements contained in vector x
+ * @tparam T_s type of element of sigma, marginal standard deviation
+ * @tparam T_l type of elements of length scale
+ *
+ * @param x std::vector of elements that can be used in stan::math::distance
+ * @param length_scale length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern32_cov(const std::vector<T_x> &x, const T_s &sigma,
+                const std::vector<T_l> &length_scale) {
+  using std::exp;
+
+  size_t x_size = x.size();
+  size_t l_size = length_scale.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern32_cov", "x", x[n]);
+
+  check_positive_finite("gp_matern32_cov", "magnitude", sigma);
+  check_positive_finite("gp_matern32_cov", "length scale", length_scale);
+
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern32_cov", "length scale", length_scale[n]);
+
+  check_size_match("gp_matern32_cov", "x dimension", x[0].size(),
```

Comment:
This will fail in the same way as the other PR that was closed. If `x` is size 0, this will seg fault.